barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/ultra_circuit_builder.cpp

```diff
@@ -2133,11 +2133,35 @@ void UltraCircuitBuilder_<Arithmetization>::create_final_sorted_RAM_gate(RamReco
     record.record_witness = this->add_variable(0);
     record.gate_index = this->num_gates;
 
+    // TODO(https://github.com/AztecProtocol/barretenberg/issues/879): this was once a single gate but has been broken
+    // into two for trace sorting.
+
+    // Create a final gate with all selectors zero; wire values are accessed by the previous RAM gate via shifted wires
+    blocks.main.populate_wires(
+        record.index_witness, record.timestamp_witness, record.value_witness, record.record_witness);
+    blocks.main.q_m().emplace_back(0);
+    blocks.main.q_1().emplace_back(0);
+    blocks.main.q_2().emplace_back(0);
+    blocks.main.q_3().emplace_back(0);
+    blocks.main.q_c().emplace_back(0);
+    blocks.main.q_arith().emplace_back(0);
+    blocks.main.q_4().emplace_back(0);
+    blocks.main.q_sort().emplace_back(0);
+    blocks.main.q_elliptic().emplace_back(0);
+    blocks.main.q_lookup_type().emplace_back(0);
+    blocks.main.q_aux().emplace_back(0);
+    if constexpr (HasAdditionalSelectors<Arithmetization>) {
+        blocks.main.pad_additional();
+    }
+    check_selector_length_consistency();
+    ++this->num_gates;
+
+    // Create an add gate ensuring the final index is consistent with the size of the RAM array
     create_big_add_gate({
         record.index_witness,
```

Comment:
Maybe it would be useful to explain how/why this big_add_gate is not related to the final_sorted_RAM_gate?  Maybe it's just me being confused tho

Reply:
Ok here's a better explanation, which could also potentially be a comment: What is now two gates (a dummy and an arithmetic) used to be a single arithmetic gate. This gate served two purposes (1) to provide wire values to the previous RAM gate via shifts, and (2) to perform a consistency check on the value in wire 1. Once we sort by gate type, the arithmetic gate will not fall immediately after the previous RAM gate, so it will not have access to the shifted wire values that it needs. Splitting the gate into two allows for the arithmetic consistency check to happen independently of the series of RAM gates that will remain sequential.

Reply:
Thanks for the explainer, I think it's worth clarifying in comments as well

Reply:
Added a similar comment